### ISSUE TITLE
[QUERY][Blob Storage] OperationContext equivalent in v12 SDK

### ISSUE THREAD

**Issue by SukruthKS** [Azure/azure-sdk-for-java#12373]
**Query/Question**
In the v12 SDK, is there something similar to `OperationContext` like in v8 SDK? I'm looking for a common way to get request details like client request Id, request Id provided by server etc. The only way I found so far is by inspecting the return value from the storage request. Is that the best way? In my application, I'm creating a common layer that all the storage requests will go through and it enables me to perform common operations like logging, metrics tracking etc. If I rely on the return values, I suppose I could use `WithResponse` version of the APIs but not all of them extend a common type (e.g., `BeginCopy` operation returns `SyncPoller`).

***Why is this not a Bug or a feature Request?***
NA

**Setup (please complete the following information if applicable):**
 - OS: Linux
 - IDE : IntelliJ
 - Version of the Library used: 12.7.0
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by rickle-msft** [comment]
Hi, @SukruthKS. Thank you for your question. You are correct that the withResponse methods are the way to gather the specified information in this version. But you're saying the different return types of some methods breaks this pattern for you?

@srnagar or @JonathanGiles can you step in either on recommended usage to satisfy the desired architecture or possibly comment on how we might make the api surface more uniform to accommodate such a pattern?

**Comment by SukruthKS** [comment]
> But you're saying the different return types of some methods breaks this pattern for you?

That's right. And also I'm not sure if any new APIs that are introduced in the future are going to follow the same convention.

**Comment by joshfree** [comment]
per offline conversation @JonathanGiles is investigating

**Comment by JonathanGiles** [comment]
@SukruthKS I'm curious how tightly correlated you need the response information to be with the client API call. For example, one potential approach would be to create a custom http policy that simply allows for you to be called back on every response. It wouldn't by default have any correlation to the specific client API call that you were performing, but I'm curious if for your use cases this approach is sufficient. If not, if you could speak to the requirements you have, we can find a suitable solution. Thanks.

**Comment by SukruthKS** [comment]
@JonathanGiles we usually want them to be correlated with the client API calls. Here is a typical scenario: we observe that a logical operation in our application was running slow or failing and we debug that it's due to the storage calls. We reach out to Azure Support and they want us to provide request IDs for them to investigate further. Also, logical operations running on multiple threads can share the same client and since there is only one instance of a custom policy in the pipeline, it can expose request IDs belonging to some other call. Having the IDs will also help us easily correlate calls with storage audit logs and our application logs.

**Comment by anuchandy** [comment]
Hi @SukruthKS, like Jonathan mentioned, you can use the custom policy to retrieve the RequestId. To correlate such RequestId to the API-call, we can use `Context` and `PipelineCallContext`.

The Context is thread-safe; API calls have an overload that takes Context param.

```java
import com.azure.core.http.HttpRequest;
import com.azure.core.http.rest.Response;
import com.azure.core.util.Context;
import com.azure.storage.blob.BlobContainerClient;
import com.azure.storage.blob.BlobContainerClientBuilder;
import com.azure.storage.blob.models.BlobContainerProperties;

import java.util.HashMap;
import java.util.Optional;

public class App {
    public static void main( String[] args ) {
        final String accountName = "<ACCOUNT-NAME>";
        final String containerName = "<CONTAINER-NAME>";
        final String containerEndpoint = String.format("https://%s.blob.core.windows.net/%s",
                accountName,
                containerName);
        final String sasToken = "SAS-TOKEN";

        BlobContainerClient blobContainerClient = new BlobContainerClientBuilder()
                .endpoint(containerEndpoint)
                .sasToken(sasToken)
                .addPolicy((callContext, next) -> {
                    Optional<Object> apiCallNameOpt = callContext.getData("api-call-name");
                    if (apiCallNameOpt.isPresent()) {
                        final String apiCallName = (String) apiCallNameOpt.get();
                        final HttpRequest request = callContext.getHttpRequest();
                        final String clientRequestId = request.getHeaders().getValue("x-ms-client-request-id");
                        System.out.println("API:" + apiCallName + " RequestId:" + clientRequestId);
                    }
                    return next.process();
                })
                .buildClient();

        Context rootContext = Context.NONE;
        // API_1
        Response<BlobContainerProperties> propertiesResponse = blobContainerClient
                .getPropertiesWithResponse(null, // leaseId
                        null, // timeout
                        rootContext.addData("api-call-name", "get-container-properties"));

        System.out.println(propertiesResponse.getStatusCode());
        // API_2
        blobContainerClient.setMetadataWithResponse(
                new HashMap<String, String>() {{ put("mdkey1", "mdval1"); }},
                null, // requestConditions
                null, // timeout
                rootContext.addData("api-call-name", "set-container-metadata"));

         System.out.println(propertiesResponse.getStatusCode());

         // OUTPUT
         //
         // API:get-container-properties RequestId:162ad5e7-691f-4df7-90ac-56b3487ed159
         // API:set-container-metadata RequestId:7df188b0-fdb5-4518-82b7-a97e49f34b2f
    }
}
```

**Comment by anuchandy** [comment]
Underneath, each `Context` instance is a node in a context-tree; when we call `addData`, it creates a new `Context` node with the given data, and the new node points back to the parent `Context`. Here the parent `Context` is the one on which we called `addData`. In this way, the `Context` tree is immutable.

So in the above sample, we start with Root context, on which we call `addData` for each of the two APIs, resulting in two new child `Context` nodes pointing to the same root, but they don't refer to each other. This way, anything specifically added to the tree for api1 is not visible|reachable to other api2, and vice-versa.

**Comment by SukruthKS** [comment]
Sorry for the delayed response, @anuchandy!

The above code still does not solve the problem of correlating our application logs with the logs emitted by the newly added pipeline policy. Our application logs are correlated based on the thread from which the logs are emitted. If the above code is to be used, it would emit the logs from a worker pool thread used by the SDK.

For e.g., let's say our application is executing a logical operation that requires uploading a file to storage. It started the execution of this operation on a thread called `thread-job-12`. It invoked the sync API of the SDK to upload the file and let's say the SDK is executing the PUT request on a thread called `workerpool-1-thread-0`. The newly added policy would emit the log line from the thread `workerpool-1-thread-0`. When querying the logs for a particular logical operation, we would get all the logs from a specific thread and the log containing the request ID would not be returned making it hard to identify the details of the storage request made by that operation.

One approach we could have taken is to get the response of an SDK synchronous call, extract the request Id header and log it on the same application thread. But not all SDK calls have the `withResponse` variation as I mentioned in the description so we cannot get the request Id.

cc @rickle-msft
